### PR TITLE
Added support for galsim_hub

### PR DESCRIPTION
**Pull request by thuiop**
After discussing a bit with @EiffL, I am taking over the implementation of galsim_hub in BTK (closes issue #72 ). I made a new implementation of the DrawBlendsGenerator to do it. This is still a work in progress ; here are the issues we are facing right now : 
- galsim_hub outputs a one band image corresponding to a COSMOS-like image. This means 1. that each time we run the model, we get a different galaxy, so by reimplementing only the render_single function, each galaxy is different in every band, which is obviously not the right behaviour ; and 2. We will face the same flux issue that we have with COSMOS.
- galsim_hub is optimized to work with batches ; this means that we have an incentive to directly generate all the galaxies we need in a batch at the same time instead of doing one at a time.
- as a generative model, we do not have a catalog. However, the model does take as input 3 parameters, which need to be generated somehow

Here are the solutions I am thinking about : 
- we can use data from the catalogs that we have to generate the input parameters ; this is especially easy to do with COSMOS since the model was trained with parameters from COSMOS to begin with. Using COSMOS does not require any change to the catalog or sampling functions. We may also imagine extracting the parameters from Catsim, or even sampling new parameters with a sampling function.
- concerning the different bands, one way would be able to use a seed which I would need to look into. A better way would be to directly reimplement the render_blend function ; this would also allow for generating all the galaxies in a blend at a time ; we may even reimplement the render_minibatch if we wish to generate all the galaxies in the batch, though that would mean more code to maintain
- the flux issue is very related to the COSMOS one ; solving the COSMOS one will probably give an answer to how to get the "right" flux ; however we should keep in mind that since the galaxies are not real ones, there is no real SED. For the time being I have adopted the solution from @mpaillassa of using the get_flux function as if we had the same magnitude in each band.

In the meantime if you want to test this feature you should install tensorflow 1.15.0 and galsim_hub (which can be done via pip) ; I have included a modified version of the tutorial in the notebooks with the GalsimHubGenerator for testing purposes.

### CODE REVIEW

**Comment by ismael-mendoza** [comment]
Thanks @thuiop for taking initiative on this, some comments 

> COSMOS since the model was trained with parameters from COSMOS to begin with.
It seems good to me to start off with just the COSMOS catalog parameters. Later we can add more complicated sampling schemes. 

> 1. that each time we run the model, we get a different galaxy, so by reimplementing only the render_single function, each galaxy is different in every band, which is obviously not the right behaviour ;

This seems related to the batch issue and how much code we want to reimplement only for `galsim_hub`. Inherently `galsim` can't do stuff in batches or even multiple bands at the same time right? So there is no way around this issue of having to do one-at-a-time if we use galsim. So I think we might need to reimplement `render_mini_batch` at least.

**Comment by thuiop** [comment]
I made a new version where I rewrite up to the render_minibatch method ; galaxies are still generated on a per blend basis ; I will try to measure the time difference when using it for the whole batch instead of for each blend.

**Review by ismael-mendoza — CHANGES_REQUESTED** [review]
Thanks for the changes @thuiop, now that I see how it would look like. It seems like there is quite a bit of repeated code! Although I think this can be salvaged with some more refactoring of the `DrawBlendsGenerator` class. But I would have to think a bit of exactly how to do this.

**Comment by thuiop** [comment]
Yes, I had to copy and paste a lot of code ; another thing is that galsim_hub only supports one band generation at the moment, but may support multiband models in the future (this is what I am working on right now) ; this will conflict with BTK multiband feature and need some more rewriting anyway. We may imagine supporting multiband catalogs in the future in which case we will have the same issue ; we need to think some more about this.

Regarding this PR, there is still a strange effect from Galsim which shows up with bright galaxies, I am currently investigating it. After that, I do not know if we should solve the redundant code problem now or in another PR.

File: btk/draw_blends.py [hunk 1]
```diff
@@ -503,3 +510,111 @@ def render_single(self, entry, filt, psf, survey):
...
+        indexes=None,
+        dim_order="NCHW",
+        galsim_hub_model="hub:Lanusse2020",
+        param_names=["flux_radius", "mag_auto", "zphot"],
+    ):
+        """Initializes the DrawBlendsGenerator class.
+
+        Args:
+            catalog (btk.catalog.Catalog) : BTK catalog object from which galaxies are taken
+            sampling_function (btk.sampling_function.SamplingFunction) : BTK sampling function
```
[the remark this answers is not in the document]
Comment (thuiop) [reply btk/draw_blends.py:544]: No, I agree, this is just a placeholder. I will write some proper documentation today.

File: btk/draw_blends.py [hunk 2]
```diff
@@ -503,3 +510,111 @@ def render_single(self, entry, filt, psf, survey):
...
+            stamp_size=stamp_size,
+            meas_bands=meas_bands,
+            multiprocessing=multiprocessing,
+            cpus=cpus,
+            verbose=verbose,
+            add_noise=add_noise,
+            shifts=shifts,
+            indexes=indexes,
+            dim_order=dim_order,
+        )
```
[the remark this answers is not in the document]
Comment (thuiop) [reply btk/draw_blends.py:578]: I do not know any but I'd be happy if someone has something like that.

**Comment by thuiop** [comment]
I think this PR is ready to be merged, I will be sending it to @EiffL for review soon.
The only remaining issue is about the dependencies. Right now galsim_hub (and tensorflow) is not a dependency of btk and we said on Wednesday that it may be better to keep it that way. But that also means that we cannot have a test for galsim_hub image generation nor have the jupyter code blocks in the documentation (I did make an independent notebook and we can switch the blocks to regular non-executable code blocks but still). Do we keep it that way ? We may also make it an "optional" dependency in a separate file (is that easy to do with poetry ?), so that we have it on the github environment but it is not mandatory for the user who does not use galsim_hub.

**Review by EiffL — COMMENTED** [review]
OMG this is seriously awesome :-D Thanks @thuiop and @ismael-mendoza , this looks really great and didn't see any obvious issues.

Just had one question about where the PSF gets applied in practice, for instance I see that in the `GalsimHubGenerator.render_single()` takes a PSF as argument, but returns draws directly the base_image in the extra_data, is it because that base_image already contains the PSF convolution?

**Comment by thuiop** [comment]
> OMG this is seriously awesome :-D Thanks @thuiop and @ismael-mendoza , this looks really great and didn't see any obvious issues.
> 
> Just had one question about where the PSF gets applied in practice, for instance I see that in the `GalsimHubGenerator.render_single()` takes a PSF as argument, but returns draws directly the base_image in the extra_data, is it because that base_image already contains the PSF convolution?

Whoops you're right, it should definitely be convolved with the PSF ! I think I removed it when I was wondering if the HST PSF was already convolved with the galaxy and forgot to add it back.

**Review by ismael-mendoza — APPROVED** [review]

**Comment by ismael-mendoza** [comment]
> OMG this is seriously awesome :-D Thanks @thuiop and @ismael-mendoza , this looks really great and didn't see any obvious issues.

Thanks for your review @EiffL and for spotting that we missed the PSF convolution!

**Pull request merged**